File: src/main/java/com/scottscmo/model/song/converters/KVMDConverter.java
package com.scottscmo.model.song.converters;

import com.scottscmo.model.song.Section;
import com.scottscmo.model.song.Song;
import com.scottscmo.util.KVMD;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public final class KVMDConverter {
    public static Song parse(String kvmdSong) {
        if (kvmdSong == null) return null;

        Map<String, Object> kvmdObj = KVMD.parse(kvmdSong);
        return new Song()
                .title(KVMD.getNamespace(kvmdObj))
                .tags((String)(KVMD.getMetadata(kvmdObj).getOrDefault("tags", "")))
                .sections(KVMD.getContent(kvmdObj))
                .order((List<String>)(KVMD.getMetadata(kvmdObj).getOrDefault("order", Collections.emptyList())));
    }

    public static String stringify(Song song) {
        return KVMD.stringify(KVMD.create(
            song.title(),
            Map.of(
                "order", song.order(),
                "tags", song.tags()
            ),
            song.sections().stream()
                    .collect(Collectors.toMap(Section::name, Section::text))
        ));
    }

    public static String stringify(Song song, List<String> textGroups, int maxLines) {
        List<Map<String, String>> transformedSectionTexts = Util.getSectionTexts(song, textGroups, maxLines);
        List<Section> transformedSections = new ArrayList<>();
        for (int i = 0; i < transformedSectionTexts.size(); i++) {
            transformedSections.add(new Section()
                    .name("s" + i)
                    .text(transformedSectionTexts.get(i)));
        }
        song.sections(transformedSections);
        song.order(transformedSections.stream().map(Section::name).toList());
        dedupeSections(song);
        return stringify(song);
    }

    private static void dedupeSections(Song song) {
        List<String> newOrder = new ArrayList<>();
        List<Section> newSections = new ArrayList<>();
        Map<String, String> visitedSections = new HashMap<>();
        for (int i = 0; i < song.sections().size(); i++) {
            Section section = song.sections().get(i);
            // use section text as key and section number as value
            String sectionKey = section.text().entrySet().stream()
                    .map(text -> "%s=%s".formatted(text.getKey(), text.getValue()))
                    .collect(Collectors.joining("&"));
            if (visitedSections.containsKey(sectionKey)) {
                // if contains, meaning it's a dupe, reuse the section number
                newOrder.add(visitedSections.get(sectionKey));
            } else {
                // unique sections
                newOrder.add(section.name());
                newSections.add(section);
                visitedSections.put(sectionKey, section.name());
            }
        }
        song.sections(newSections);
        song.order(newOrder);
    }
}

File: src/main/kotlin/com/scottscmo/model/song/converters/KVMDConverter.kt
package com.scottscmo.model.song.converters

import com.scottscmo.model.song.Section
import com.scottscmo.model.song.Song
import com.scottscmo.util.KVMD

object KVMDConverter {
    fun parse(kvmdSong: String?): Song? {
        if (kvmdSong == null) return null

        val kvmdObj = KVMD.parse(kvmdSong)
        return Song().apply {
            title = KVMD.getNamespace(kvmdObj)
            tags = KVMD.getMetadata(kvmdObj)["tags"] as String ?: ""
            order = KVMD.getMetadata(kvmdObj)["order"] as List<String>
            sections = KVMD.getContent(kvmdObj).entries
                .sortedBy { it.key }
                .map {
                    Section().apply {
                        name = it.key
                        text = it.value as Map<String, String>
                    }
                }.toList()
        }
    }

    fun stringify(song: Song): String {
        return KVMD.stringify(KVMD.create(
            song.title,
            mapOf(
                "order" to song.order,
                "tags" to song.tags
            ),
            song.sections.associate { it.name to it.text }
        ))
    }

    fun stringify(song: Song, textGroups: List<String>, maxLines: Int): String {
        song.sections = Util.getSectionTexts(song, textGroups, maxLines).mapIndexed { index, sectionText ->
            Section().apply {
                name = "s$index"
                text = sectionText
            }
        }
        song.order = song.sections.map { it.name }
        dedupeSections(song)
        return stringify(song)
    }

    private fun dedupeSections(song: Song) {
        val newOrder = mutableListOf<String>()
        val newSections = mutableListOf<Section>()
        val visitedSections = mutableMapOf<String, String>()
        for (i in song.sections.indices) {
            val section = song.sections[i]
            // use section text as key and section number as value
            val sectionKey = section.text.entries.joinToString("&") { "${it.key}=${it.value}" }
            if (visitedSections.containsKey(sectionKey)) {
                // if contains, meaning it's a dupe, reuse the section number
                newOrder.add(visitedSections[sectionKey]!!)
            } else {
                // unique sections
                newOrder.add(section.name)
                newSections.add(section)
                visitedSections[sectionKey] = section.name
            }
        }
        song.sections = newSections
        song.order = newOrder
    }
}
